DR-2112 Request smaller scope to verify user has access to the google billing account

nmalfroy opened the pull request:
This changes the required scope for billing account creation from
`https://www.googleapis.com/auth/cloud-platform`
to
`https://www.googleapis.com/auth/cloud-billing.readonly`

nmalfroy committed:
Request smaller scope to verify user has access to the google billing account

nmalfroy committed:
Use an even smaller scope

nmalfroy committed:
Use the broader scope for tests

At some point the users should be granted the narrower scopes

pshapiro4broad commented on datarepo-clienttests/src/main/java/common/utils/AuthenticationUtils.java:
```diff
@@ -29,7 +29,7 @@ private AuthenticationUtils() {}
   // what the UI requests, so our tests represent actual user behavior
   private static final List<String> userLoginScopes =
-      Arrays.asList("openid", "email", "profile", "https://www.googleapis.com/auth/cloud-platform");
+      Arrays.asList("openid", "email", "profile", "https://www.googleapis.com/auth/cloud-billing");
```
May as well update to `List.of()` while you're here
```suggestion
      List.of("openid", "email", "profile", "https://www.googleapis.com/auth/cloud-billing");
```

pshapiro4broad commented on src/main/java/bio/terra/service/profile/google/GoogleBillingService.java:
```diff
@@ -43,7 +43,7 @@ private static CloudBillingClient accessClient(AuthenticatedUserRequest user) {
       // for more information.
       List<String> scopes =
-          Collections.singletonList("https://www.googleapis.com/auth/cloud-platform");
+          Collections.singletonList("https://www.googleapis.com/auth/cloud-billing.readonly");
```
```suggestion
          List.of("https://www.googleapis.com/auth/cloud-billing.readonly");
```
Also, since it's only used inside the `if`, it would be better to declare it inside the `if`.

pshapiro4broad commented on src/test/java/bio/terra/common/auth/AuthService.java:
```diff
@@ -24,7 +24,7 @@
   // behavior:
   private List<String> userLoginScopes =
-      Arrays.asList("openid", "email", "profile", "https://www.googleapis.com/auth/cloud-platform");
+      Arrays.asList("openid", "email", "profile", "https://www.googleapis.com/auth/cloud-billing");
```
If this has the same meaning as `AuthenticationUtils.userLoginScopes` it would be better to make that constant public and use it here instead of duplicating it.

pshapiro4broad commented on src/main/java/bio/terra/service/profile/google/GoogleBillingService.java:
```diff
@@ -43,7 +43,7 @@ private static CloudBillingClient accessClient(AuthenticatedUserRequest user) {
       List<String> scopes =
-          Collections.singletonList("https://www.googleapis.com/auth/cloud-platform");
+          Collections.singletonList("https://www.googleapis.com/auth/cloud-billing.readonly");
       if (serviceAccountCredentials.createScopedRequired()) {
```
There are other cases in `GoogleProjectService` where the code checks `createScopeRequired()` and then uses `https://www.googleapis.com/auth/cloud-platform`. Do those scope not need to be narrowed as well?

nmalfroy replied:
In that case, those are scope that we can explicitly ask for from the service account so I'm not as worried there

pshapiro4broad reviewed (approved):
It looks like the scopes are duplicated in the openapi yaml and in our code. It would be nice if there was a way to avoid this duplication and have the code get it from the yaml too, not sure what a good way to do that would be. I know you can use reflection to read the annotation settings but that seems a bit awkward.

nmalfroy committed:
Increase scope for when we create a token for the service account

Since it needs more permissions

tlangs commented on src/test/java/bio/terra/common/auth/AuthService.java:
```diff
@@ -37,9 +36,9 @@
   // behavior:
   private List<String> userLoginScopes =
-      Arrays.asList("openid", "email", "profile", "https://www.googleapis.com/auth/cloud-platform");
+      List.of("openid", "email", "profile", "https://www.googleapis.com/auth/cloud-billing");
```
whats the reason for this being `cloud-billing` and not `cloud-billing.readonly`?

nmalfroy replied:
I may need help from DevOps to update the dumbledore user to allow that scope. It's very strange.  For what it's worth, testing manually works

nmalfroy replied:
Ok, I reduced the scopes requested by the tests so we should be all set now

nmalfroy commented:
Great question!  That appears to be because the user needs the explicit
readonly scope and currently doesn't have it. I'm gonna see if I can figure
out how to do it

On Wed, Sep 22, 2021 at 4:24 PM Trevyn Langsford ***@***.***>
wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In src/test/java/bio/terra/common/auth/AuthService.java
> <https://github.com/DataBiosphere/jade-data-repo/pull/1066#discussion_r714283058>
> :
>
> > @@ -37,9 +36,9 @@
>    // this should always match exactly what the UI requests, so our tests represent actual user
>    // behavior:
>    private List<String> userLoginScopes =
> -      Arrays.asList("openid", "email", "profile", "https://www.googleapis.com/auth/cloud-platform");
> +      List.of("openid", "email", "profile", "https://www.googleapis.com/auth/cloud-billing");
>
> whats the reason for this being cloud-billing and not
> cloud-billing.readonly?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/DataBiosphere/jade-data-repo/pull/1066#pullrequestreview-761370394>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABK7N63GDCBPUHR4KCYXZR3UDI3PTANCNFSM5EMUEM3A>
> .
>

nmalfroy committed:
Lower the required scopes for the tests

Now that I've set the domain-wide delegate options to include that the https://www.googleapis.com/auth/cloud-billing.readonly scope
